### DISCUSSION TITLE
smartdns的dot协议853端口应该用什么证书呀？小米手机连上就报错

### DISCUSSION THREAD

**Discussion by 1289491916** [pymumu/smartdns#1456] [Q&A]
这是服务器上的报错，说是TLS握手失败，我其实刚刚开始部署手机还可以连上，用了一分多钟就不行了，这是我证书用错了么
![image](https://github.com/pymumu/smartdns/assets/51084029/f151d01d-5ea0-47dc-ac52-ce4f49f640f7)

下面是我的证书格式：下载的是腾讯云的免费证书的other那一个版本，没用nginx和apache那些，是不是用错了？
smart.xxx.com_bundle.pem
![image](https://github.com/pymumu/smartdns/assets/51084029/5149f4f5-78db-44ac-bc8b-3cf1ff87befb)
smart.xxxx.com.key
![image](https://github.com/pymumu/smartdns/assets/51084029/f12519f7-9c4e-4830-8fa3-379108e9d591)
查了百度也没说过DOT协议用哪种类型的证书，而且smartdns服务启动也没报错，就是懵的，麻烦大佬们帮忙解答一下吧

**Comment by PikuZheng** [comment]
自架上游吗？是自签证书还是向可信ca申请的？
证书用的和https一样的。如果是刚开始好用过了一会儿不好用，可能是有中间人攻击

**Reply by 1289491916** [reply]
下载的是腾讯云的免费证书的other那一个版本，没用nginx和apache那些

**Reply by PikuZheng** [reply]
你是说给smartdns指派了一个证书，smartdns作为tls服务器？
smartdns需要两个文件，ca证书和域名证书合并到一起作为证书文件（一般叫fullchain.pem；key文件作为证书密钥。

**Reply by 1289491916** [reply]
我想请教一下如何合并呢？我应该下载哪个版本的腾讯云证书，用哪个文件的内容进行合并呢？
证书版本我虽然分得清也经常部署，但是涉及到证书内容组成我就是彻底的小白了，目前证书文件是-----BEGIN CERTIFICATE-----的内容，但是不清楚CA证书是从哪里下载，是从腾讯云下载根证书还是去亚洲诚信官网下载“TrustAsia EV TLS RSA CA”证书呢？
是直接拼接文本进行合并还是用openssl之类的工具进行计算合并呢？
我的疑问可能有点多🙏，如果您愿意分享您的知识我会十分感激！

**Reply by PikuZheng** [reply]
![image](https://github.com/pymumu/smartdns/assets/12579522/1a7cdefa-facd-4ffd-8bea-5e21ddbfef8f)
刚随便申请了一下。“nginx”和“其他”是一样的。使用里面的` _bundle.pem`，这个是链好的。还有`.key`，只需要这两个文件。仔细看了一下你用的是对的（一开始我用手机回复的没看清楚）

**Reply by PikuZheng** [reply]
手机上的安全dns，需要用域名连接（因为申请的证书里没有ip）。还有就是，如果你的tls服务器部署在国外，可能会被打（853端口概率被墙

**Reply by 1289491916** [reply]
遗憾了，买的是阿里云的ECS服务器，成都区的，不会是运营商搞我吧

**Comment by pymumu** [comment]
和客户端握手失败了吧。

改一下代码，打印下错误码看看：

dns_server.c的6176行
```
		ret = _ssl_do_accept(tls_client);
		if (ret == 0) {
			goto errout;
		} else if (ret < 0) {
```

改成
```
		ret = _ssl_do_accept(tls_client);
		if (ret <= 0) {

```

**Reply by 1289491916** [reply]
修改并且编译了，好像打印结果没啥变化🤔
![image](https://github.com/pymumu/smartdns/assets/51084029/696eb820-1c9e-4c80-add3-5e1e04da24d9)


以下是修改的代码，是否正确：
![image](https://github.com/pymumu/smartdns/assets/51084029/372110f0-ea34-40cc-8390-2123012ae9c9)

**Comment by pymumu** [comment]
改的没有问题，确认是否保存和编译了吧。看log行号不对。

**Reply by 1289491916** [reply]
行号不对，在6252行去了，
![image](https://github.com/pymumu/smartdns/assets/51084029/cd1af2b7-710a-4436-8a30-a576f1aa39bf)
